vault/keysource.go

```diff
@@ -0,0 +1,241 @@
+	BackendPath  string
+	VaultAddress string
+	CreationDate time.Time
+}
+
+// NewMasterKeysFromURIs gets lots of keys from lots of URIs
+func NewMasterKeysFromURIs(uris string) ([]*MasterKey, error) {
+	var keys []*MasterKey
+	if uris == "" {
+		return keys, nil
+	}
+	uriList := strings.Split(uris, ",")
+	for _, uri := range uriList {
+		if uri == "" {
+			continue
+		}
+		key, err := NewMasterKeyFromURI(uri)
+		if err != nil {
+			return nil, err
+		}
+		keys = append(keys, key)
+	}
+	return keys, nil
+}
+
+// NewMasterKeyFromURI obtains the vaultAddress the transit backend path and the key name from the full URI of the key
+func NewMasterKeyFromURI(uri string) (*MasterKey, error) {
+	log.Debugln("Called NewMasterKeyFromURI with uri: ", uri)
+	var key *MasterKey
+	if uri == "" {
+		return key, nil
+	}
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	if u.Scheme == "" {
+		return nil, fmt.Errorf("missing scheme in vault URL (should be like this: https://vault.example.com:8200/v1/transit/keys/keyName, got: %v", uri)
+	}
+	backendPath, keyName, err := getBackendAndKeyFromPath(u.EscapedPath())
+	if err != nil {
+		return nil, err
+	}
+	u.Path = ""
+	return NewMasterKey(u.String(), backendPath, keyName), nil
+
+}
+
+func getBackendAndKeyFromPath(path string) (string, string, error) {
+	path = strings.TrimPrefix(path, "/")
+	path = strings.TrimSuffix(path, "/")
+	values := strings.Split(path, "/")
+	// minimum lenght should be 4 "v1", "transit", "keys", "keyName"
+	if len(values) < 4 {
+		return "", "", fmt.Errorf("The path to the key is not long enough: (eg. https://vault.example.com:8200/v1/transit/keys/keyName")
+	}
+	log.Debugf("Path: %v \nValues: %v\n", path, values)
+	if values[0] != "v1" {
+		return "", "", fmt.Errorf("probably forgot v1 in the URI")
+	}
+	if values[len(values)-2] != "keys" {
+		return "", "", fmt.Errorf("probably forgot 'keys' in the URI")
+	}
+	return filepath.Join(values[1 : len(values)-2]...), values[len(values)-1], nil
+}
+
+// NewMasterKey creates a new MasterKey from a vault address, transit backend path and a key name and setting the creation date to the current date
+func NewMasterKey(addess, backendPath, keyName string) *MasterKey {
+	mk := &MasterKey{
+		VaultAddress: addess,
+		BackendPath:  backendPath,
+		KeyName:      keyName,
+		CreationDate: time.Now().UTC(),
+	}
+	log.Debugln("Created Vault Master Key: ", mk)
+	return mk
+}
+
+// EncryptedDataKey returns the encrypted data key this master key holds
+func (key *MasterKey) EncryptedDataKey() []byte {
+	return []byte(key.EncryptedKey)
+}
+
+// SetEncryptedDataKey sets the encrypted data key for this master key
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Encrypt takes a sops data key, encrypts it with Vault Transit and stores the result in the EncryptedKey field
+func (key *MasterKey) Encrypt(dataKey []byte) error {
+	path := filepath.Join(key.BackendPath, "encrypt", key.KeyName)
```

Comment:
Same as above, probably want
```suggestion
	path := path.Join(key.BackendPath, "encrypt", key.KeyName)
```